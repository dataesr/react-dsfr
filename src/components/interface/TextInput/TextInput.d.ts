import * as React from 'react';

export type TextInputHint = string | Object | any[];

export type TextInputMessageType = "error" | "valid" | "";

export type TextInputType = "text" | "date" | "number" | "email" | "password";

export type TextInputClassName = string | Object | any[];

export interface TextInputProps extends React.InputHTMLAttributes<HTMLInputElement> {
    textarea?: boolean;
    type?: TextInputType;
    label?: string;
    hint?: TextInputHint;
    messageType?: TextInputMessageType;
    message?: string;
    required?: boolean;
    className?: TextInputClassName;
    withAutoValidation?: boolean;
    onBlur?: (...args: any[]) => any;
}

declare const TextInput: React.FC<TextInputProps>;

export default TextInput;

