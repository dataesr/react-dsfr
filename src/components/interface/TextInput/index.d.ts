import type TextInput from './TextInput';

export type { TextInput };
