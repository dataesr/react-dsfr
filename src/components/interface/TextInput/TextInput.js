import React, { forwardRef } from 'react';

import PropTypes from 'prop-types';
import classNames from 'classnames';
import { v4 as uuidv4 } from 'uuid';

/**
 *
 * @visibleName TextInput
 */
const TextInput = forwardRef((props, ref) => {
  const {
    textarea,
    inactive,
    label,
    hint,
    message,
    value,
    onChange,
    messageType,
    className,
  } = props;

  const _classNameWrapper = classNames('fr-input-group', {
    [`fr-input-group--${messageType}`]: messageType,
  }, className);
  const _className = classNames('fr-input', {
    [`fr-input--${messageType}`]: messageType,
  });

  const inputId = uuidv4();
  const hintId = hint && uuidv4();
  const messageId = message && uuidv4();
  return (
    <div className={_classNameWrapper}>
      {label && <label className="fr-label" htmlFor={inputId} aria-describedby={hintId || messageId || undefined}>{label}</label>}
      {hint && <p className="fr-hint-text" id={hintId}>{hint}</p>}
      {
        (textarea)
          ? (
            <textarea
              ref={ref}
              className={_className}
              disabled={inactive}
              id={inputId}
              value={value}
              onChange={onChange}
            />
          )
          : (
            <input
              ref={ref}
              className={_className}
              disabled={inactive}
              type="text"
              id={inputId}
              value={value}
              onChange={onChange}
            />
          )
    }
      {(message && messageType === 'error') && <p id={messageId} className="fr-error-text">{message}</p>}
      {(message && messageType === 'valid') && <p id={messageId} className="fr-valid-text">{message}</p>}
    </div>
  );
});
TextInput.defaultProps = {
  textarea: false,
  inactive: false,
  hint: '',
  onChange: () => {},
  messageType: '',
  message: '',
  label: null,
  className: '',
};
TextInput.propTypes = {
  textarea: PropTypes.bool,
  inactive: PropTypes.bool,
  label: PropTypes.string,
  hint: PropTypes.string,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  messageType: PropTypes.oneOf(['error', 'valid', '']),
  message: PropTypes.string,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
};

export default TextInput;
