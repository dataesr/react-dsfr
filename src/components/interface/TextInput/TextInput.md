## Champs de saisie simple - Input 

Le champ simple est un champ de saisie libre, qui accepte une courte ligne de contenu (texte ou/ et nombre). Le libellé se trouve au-dessus du champs de saisie, pour faciliter la lecture.

### État par défaut

```jsx
import { useState } from 'react';

const [value, setValue] = useState('');

<TextInput
  value={value}
  onChange={e => setValue(e.target.value)}
  label="Text input du gouvernement"
/>
```

### État validé et erreur

Ils sont signalés par le changement de couleur de la bordure (cf. couleurs fonctionnelles 🔗 ) et l’apparition d’un message sous le champs, obligatoire pour le cas d’erreur.
Le message et le label sont liés par leurs attribut `id` et `aria-describedby` qui sont gérés par le composant.

```jsx
import { useState } from 'react';

const [value, setValue] = useState('');

<>
  <TextInput
    value={value}
    onChange={e => setValue(e.target.value)}
    label="Text input du gouvernement"
    messageType="error"
    message="Je suis un message ! "
  />

  <TextInput
    value={value}
    onChange={e => setValue(e.target.value)}
    label="Text input du gouvernement"
    messageType="valid"
    message="Je suis un message ! "
  />
</>
```

### État inactif

L'état inactif est utilisé lorsque le champ ne peut être utilisé. Il indique que utilisateur ne peux pas saisir de contenus, par exemple jusqu'à ce qu'une autre action soit terminée. Il ne doit être utilisé que très ponctuellement, préférez masquer le champs si celui-ci n’est pas requis.

```jsx
import { useState } from 'react';

const [value, setValue] = useState('');

<TextInput
  inactive
  value={value}
  onChange={e => setValue(e.target.value)}
  label="Text input du gouvernement"
/>
```

### Champ avec texte d’aide
Il est recommandé d’ajouter un texte d’aide sous le libellé afin de faciliter la saisie

```jsx
import { useState } from 'react';

const [value, setValue] = useState('');

<TextInput
  value={value}
  hint="Le label du dessus n'est pas assez explicite ?"
  onChange={e => setValue(e.target.value)}
  label="Text input du gouvernement"
/>
```

## Zone de texte - Textarea
Le champs “zone de texte” est un champ de saisie libre, qui accepte plus d’une ligne de contenu (texte ou/ et nombre). Il reprend le style du champ simple et augmente uniquement sa hauteur.

```jsx
import { useState } from 'react';

const [value, setValue] = useState('');

<TextInput
  textarea
  value={value}
  hint="Le label du dessus n'est pas assez explicite ?"
  onChange={e => setValue(e.target.value)}
  label="Text input du gouvernement"
/>
```
