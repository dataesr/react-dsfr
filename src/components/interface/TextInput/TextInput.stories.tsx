import { ComponentMeta, ComponentStory } from '@storybook/react';

import TextInput from './TextInput';

export default {
  title: 'TextInput',
  component: TextInput,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} as ComponentMeta<typeof TextInput>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof TextInput> = (args) => <TextInput {...args} />;

export const Text = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Text.args = {};

export const TextArea = Template.bind({});
TextArea.args = {
  textarea: true,
  rows: 3,
};
