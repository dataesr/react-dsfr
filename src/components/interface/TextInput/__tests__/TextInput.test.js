import '@testing-library/jest-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { v4 as uuidv4 } from 'uuid';
import TextInput from '../TextInput';

jest.mock('uuid', () => ({
  v4: jest.fn(),
}));

Enzyme.configure({ adapter: new Adapter() });

const onChangeSpy = jest.fn();
const initialProps = {
  value: 'init-value',
  label: 'label',
  onChange: onChangeSpy,
};

describe('<TextInput />', () => {
  let wrapper;
  const event = {
    target: {
      value: 'New value',
    },
  };

  beforeEach(() => {
    uuidv4.mockImplementationOnce(() => 'xxxxxxx');
    wrapper = (props = {}) => shallow(
      <TextInput {...props} />,
    );
  });

  it('renders correctly', () => {
    const component = renderer
      .create(<TextInput {...initialProps} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('props value of TextInput', () => {
    const props = {
      value: 'test-value',
      label: 'test-label',
      onChange: onChangeSpy,
    };
    const component = wrapper(props);
    expect(component.find('input').prop('value')).toBe('test-value');
    expect(component.find('label').text()).toBe('test-label');
  });

  test('onChange of TextInput', () => {
    const component = wrapper(initialProps);
    component.find('input').simulate('change', event);
    expect(onChangeSpy).toHaveBeenCalledWith(event);
  });
});
