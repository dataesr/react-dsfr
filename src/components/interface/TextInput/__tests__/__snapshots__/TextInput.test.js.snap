// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> renders correctly 1`] = `
<div
  className="fr-input-group"
  data-testid="textinput"
>
  <label
    className="fr-label"
    htmlFor="xxxxxxx"
  >
    label
  </label>
  <input
    aria-describedby=""
    className="fr-input"
    id="xxxxxxx"
    onBlur={[Function]}
    onChange={[MockFunction]}
    required={false}
    type="text"
    value="init-value"
  />
</div>
`;
